### PR TITLE
MRG: Simulation eval

### PR DESCRIPTION
**Pull request by Eric89GXL**
Taking over #2295.

### CODE REVIEW

**Comment by agramfort** [comment]
did you cherry pick my commit?

**Comment by Eric89GXL** [comment]
Nope, forgot. I'll add it.

**Comment by agramfort** [comment]
thanks. Let me know when I shall review.

**Comment by Eric89GXL** [comment]
Okay @agramfort ready for a round of review. FWIW I think we need to have better tests for the cHPI, but I want to do that as a subsequent PR after this one is merged. I think it's my highest priority right now actually, so I can probably do it in a few days. If we can't get it actually validated before release, we can just remove the `chpi` option for now.

**Comment by agramfort** [comment]
travis complains

File: examples/simulation/plot_simulate_raw_data.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
+"""
+==============================
+Generate simulated raw data
+==============================
+
```
Comment (agramfort) [inline examples/simulation/plot_simulate_raw_data.py]: can you add a description of what the example does?

File: examples/simulation/plot_simulate_raw_data.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+##############################################################################
+# Simulate raw data
+raw_sim = simulate_raw(raw, stc, trans_fname, src, bem_fname,
+                       ecg=True, blink=True, n_jobs=2, verbose=True)
+raw_sim.plot()
+
+##############################################################################
+# Plot evoked data
+events = find_events(raw_sim)
+events[:, 2] = 1
```
Comment (agramfort) [inline examples/simulation/plot_simulate_raw_data.py]: do you need this line?

File: examples/simulation/plot_simulate_raw_data.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+bem_fname = (data_path +
+             '/subjects/sample/bem/sample-5120-5120-5120-bem-sol.fif')
+
+# Load real data as the template
+raw = Raw(raw_fname).crop(0., 60., copy=False)  # 60 sec is enough
+
+##############################################################################
+# Generate dipole time series
+n_dipoles = 4  # number of dipoles to create
+dur = 2.  # duration of each epoch/event
```
Comment (agramfort) [inline examples/simulation/plot_simulate_raw_data.py]: dur -> duration or better epoch_duration

File: examples/simulation/plot_simulate_raw_data.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+fig, ax = plt.subplots(1)
+ax.plot(times, 1e9 * stc.data.T)
+ax.set(ylabel='Amplitude (nAm)', xlabel='Time (sec)')
+fig.show()
+
+##############################################################################
+# Simulate raw data
+raw_sim = simulate_raw(raw, stc, trans_fname, src, bem_fname,
+                       ecg=True, blink=True, n_jobs=2, verbose=True)
+raw_sim.plot()
```
Comment (agramfort) [inline examples/simulation/plot_simulate_raw_data.py]: magnetometers look particularly noisy. Any obvious reason?

Comment (agramfort) [inline examples/simulation/plot_simulate_raw_data.py]: also the SNR parameter via the noise cov is implicit. I would make it explicit in the example.

Comment (Eric89GXL) [inline examples/simulation/plot_simulate_raw_data.py]: there is no snr parameter currently, because it doesn't really make sense with head movements, see notes of `simulate_raw`

File: examples/simulation/plot_simulate_raw_data.py (hunk 2 above)
Comment (Eric89GXL) [inline examples/simulation/plot_simulate_raw_data.py]: in general yes if there are head movements, but not in this particular example so I'll remove it

File: mne/simulation/metrics.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+def source_estimate_quantification(stc1, stc2, metric='rms'):
+    """Helper function to calculate matrix similarities.
+
+    Parameters
+    ----------
+    stc1 : SourceEstimate
+        First source estimate for comparison.
+    stc2 : SourceEstimate
+        First source estimate for comparison.
```
Comment (agramfort) [inline mne/simulation/metrics.py]: First -> Second

File: mne/simulation/metrics.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    # Calculate root mean square difference between two matrices
+    if metric == 'rms':
+        return np.sqrt(np.mean((data1 - data2) ** 2))
+
+    # Calculate correlation coefficient between matrix elements
+    elif metric == 'cosine':
+        score = (np.correlate(data1.flatten(), data2.flatten()) /
+                 (norm(data1) * norm(data2)))
+        return 1 - score
```
Comment (agramfort) [inline mne/simulation/metrics.py]: 1 -> 1.

I would just be explicit

score = np.dot(data1.flatten(), data2.flatten()) / (norm(data1) * norm(data2))

File: mne/simulation/raw.py [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+                 blink=False, ecg=False, chpi=False, head_pos=None,
+                 mindist=1.0, interp='cos2', n_jobs=1, random_state=None,
+                 verbose=None):
+    """Simulate raw data with head movements
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The raw template to use for simulation. The ``info``, ``times``,
+        and potentially ``first_samp`` properties will be used.
```
Comment (agramfort) [inline mne/simulation/raw.py]: I would pass an info not raw as for evoked. You have the times array from stc.times + you can add a single param for first_samp is needed.

File: mne/simulation/raw.py [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+    bem : str | dict
+        BEM solution  corresponding to the stc. If string, should be a BEM
+        solution filename (e.g., "sample-5120-5120-5120-bem-sol.fif").
+    cov : instance of Covariance | str | None
+        The sensor covariance matrix used to generate noise. If None,
+        no noise will be added. If 'simple', a basic (diagonal) ad-hoc
+        noise covariance will be used. If a string, then the covariance
+        will be loaded.
+    blink : bool
+        If true, add simulated blink artifacts.
```
Comment (agramfort) [inline mne/simulation/raw.py]: can you document how it is done?

File: mne/simulation/raw.py [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+    ecg : bool
+        If true, add simulated ECG artifacts.
```
Comment (agramfort) [inline mne/simulation/raw.py]: same here can you document how it is done?

File: mne/simulation/raw.py (hunk 7 above)
Comment (Eric89GXL) [inline mne/simulation/raw.py]: You'd need to pass 1) `info`, 2) `times`, 3) `first_samp`, and 4) replicate `raw.time_as_index` to get the same functionality as now, so I'm not sure it's worth it. As it is now, we can eventually also make it so `head_pos='estimated'` where it estimates the head positions from raw using cHPI information, whereas by not passing `raw` we'd have to do it in separate steps.

File: examples/simulation/plot_simulate_raw_data.py (hunk 4 above)
Comment (Eric89GXL) [inline examples/simulation/plot_simulate_raw_data.py]: I'm not sure why the magnetometers are so bad actually, in the `raw` data plot I can even clearly see the activations in the magnetometers. I wonder if it has something to do with the whitening, I'll investigate further.

File: mne/simulation/raw.py [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+            0.2 * np.hanning(int(0.26 * info['sfreq']))], axis=-1)
+        ecg_data = np.convolve(cardiac_data, cardiac_kernel,
+                               'same')[np.newaxis, :] * 15e-8
+
+        # Add rescaled noisy data to ECG ch
+        ch = pick_types(info, meg=False, eeg=False, ecg=True)
+        noise = rng.randn(ecg_data.shape[1]) * 1.5e-5
+        if len(ch) >= 1:
+            raw_data[ch[-1], :] = ecg_data * 2e3 + noise
+        del cardiac_data, cardiac_kernel, max_beats, cardiac_idx
```
Comment (agramfort) [inline mne/simulation/raw.py]: the blinck + ecg generation seems a bit magic to me.

File: mne/simulation/raw.py [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+        if interp == 'zero':
+            interps = None
+        else:
+            if interp == 'linear':
+                interps = np.linspace(1, 0, n_pts, endpoint=False)
+            else:  # interp == 'cos2':
+                interps = np.cos(0.5 * np.pi * np.arange(n_pts)) ** 2
+            interps = np.array([interps, 1 - interps])
+
+        assert not used[offsets[fi - 1]:offsets[fi]].any()
```
Comment (agramfort) [inline mne/simulation/raw.py]: do you mean all() ?

File: mne/simulation/raw.py (hunk 10 above)
Comment (Eric89GXL) [inline mne/simulation/raw.py]: I'll add comments

File: mne/simulation/raw.py (hunk 11 above)
Comment (Eric89GXL) [inline mne/simulation/raw.py]: no, this ensures that the points that are about to be written/saved/generated have not been done yet, the `.all()` comes later

File: mne/simulation/raw.py [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+                    sinusoids[fidx] = hpi_mag * np.sin(sinusoids[fidx])
+                raw_data[meg_picks, time_sl] += \
+                    _interp(last_fwd_chpi, fwd_chpi, sinusoids, this_interp)
+
+        assert used[offsets[fi - 1]:offsets[fi]].all()
+
+        # prepare for next iteration
+        last_fwd, last_fwd_blink, last_fwd_ecg, last_fwd_chpi = \
+            fwd, fwd_blink, fwd_ecg, fwd_chpi
+    assert used.all()
```
Comment (Eric89GXL) [inline mne/simulation/raw.py]: ... here

File: examples/simulation/plot_simulate_raw_data.py (hunk 4 above)
Comment (Eric89GXL) [inline examples/simulation/plot_simulate_raw_data.py]: I was computing the noise cov using the entire epoch instead of the baseline, whoops!

**Comment by Eric89GXL** [comment]
@agramfort comments thus far addressed (see last commit) and the CIs are happy.

File: mne/simulation/raw.py (hunk 7 above)
Comment (agramfort) [inline mne/simulation/raw.py]: ok point taken

File: mne/io/__init__.py [hunk 13]
```diff
@@ -5,6 +5,7 @@
 #
 # License: BSD (3-clause)
 
+from .constants import FIFF
```
Comment (agramfort) [inline mne/io/__init__.py]: why did you need this?

I prefer avoiding crowding too much the module levels (yes I know we already do it a lot...)

File: mne/io/__init__.py [hunk 14]
```diff
@@ -21,6 +22,7 @@
 from . import pick
 
 from .array import RawArray
+from .base import _BaseRaw
```
Comment (agramfort) [inline mne/io/__init__.py]: same here

File: mne/simulation/raw.py [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+           the left (-x) of head center and three head radii below (+z)
+           head center.
+        4. Activations only affect MEG channels.
+
+    .. versionadded:: 0.10.0
+
+    References
+    ----------
+    .. [1] Bentivoglio et al. "Analysis of blink rate patterns in normal
+           subjects" Movment Disorders, 1997 Nov;12(6):1028-34.
```
Comment (agramfort) [inline mne/simulation/raw.py]: Movment -> Movement

File: mne/simulation/tests/test_metrics.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    assert_true(E2_rms == 0.)
+    assert_almost_equal(E1_cos, 0.)
+    assert_almost_equal(E2_cos, 0.)
+    stc_bad = stc2.copy().crop(0, 1)
+    assert_raises(ValueError, source_estimate_quantification, stc1, stc_bad)
+    stc_bad = stc2.copy()
+    stc_bad.times -= 0.1
+    assert_raises(ValueError, source_estimate_quantification, stc1, stc_bad)
+    assert_raises(ValueError, source_estimate_quantification, stc1, stc2,
+                  metric='foo')
```
Comment (agramfort) [inline mne/simulation/tests/test_metrics.py]: this test is slow. I would just test the source_estimate_quantification with simulated stcs that are surface or volume if needed. It's not using at all the fact that dSPM or MNE was used.

File: mne/simulation/tests/test_raw.py [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+    assert_raises(ValueError, simulate_raw, raw, stc, trans, src, sphere,
+                  interp='foo')
+    assert_raises(TypeError, simulate_raw, raw, stc, trans, src, sphere,
+                  head_pos=1.)
+    assert_raises(RuntimeError, simulate_raw, raw, stc, trans, src, sphere,
+                  head_pos=hp_fname)  # ends up with t>t_end
+    head_pos_sim_err = deepcopy(head_pos_sim)
+    head_pos_sim_err[-1.] = head_pos_sim_err[1.]  # negative time
+    assert_raises(RuntimeError, simulate_raw, raw, stc, trans, src, sphere,
+                  head_pos=head_pos_sim_err)
```
Comment (agramfort) [inline mne/simulation/tests/test_raw.py]: this test takes > 15s on my box. Any idea why? can we improve?

File: mne/simulation/tests/test_source.py [hunk 18]
```diff
@@ -44,7 +45,8 @@ def test_generate_stc():
     tstep = 1e-3
 
     stc_data = np.ones((len(labels), n_times))
-    stc = generate_stc(fwd['src'], mylabels, stc_data, tmin, tstep)
+    with warnings.catch_warnings(record=True):
+        stc = generate_stc(fwd['src'], mylabels, stc_data, tmin, tstep)
```
Comment (agramfort) [inline mne/simulation/tests/test_source.py]: why still testing the old functions?

**Comment by agramfort** [comment]
doing raw_sim.plot() leads to rather weak signals. It seems there is really little brain noise / signal besides the sinusoids introduced. Any idea why? Is it because of the color of the brain noise you don't take into account?

**Comment by Eric89GXL** [comment]
Yeah, the `cov='simple'` is a diagonal noise cov with not large amplitudes (from `mne.make_ad_hoc_cov`, which got values from MNE-C). I can put the measured noise covariance back in, that one is much noisier.

File: mne/io/__init__.py (hunk 14 above)
Comment (Eric89GXL) [inline mne/io/__init__.py]: This one shouldn't matter, though, because users won't see it (private)

File: mne/io/__init__.py (hunk 13 above)
Comment (Eric89GXL) [inline mne/io/__init__.py]: it simplifies the import process in other places, but I can remove this one

File: mne/simulation/tests/test_source.py (hunk 18 above)
Comment (Eric89GXL) [inline mne/simulation/tests/test_source.py]: Looking at it, it's not as trivial to get this functionality as it was with `generate_stc` because of the forced use of `data_fun`. There used to be both `generate_stc` (which you could pass `data` to) and `generate_sparse_stc`, now there is only `simulate_sparse_stc`. Should there also be `simulate_stc` that takes a `data` parameter? I wasn't around for these discussions so I'm not sure.

File: mne/simulation/tests/test_raw.py (hunk 17 above)
Comment (Eric89GXL) [inline mne/simulation/tests/test_raw.py]: I'm cutting the number of channels down for a ~3x speed increase

**Comment by Eric89GXL** [comment]
Okay @agramfort comments addressed again.

**Comment by agramfort** [comment]
ok LGTM

@yousrabk @wronk do you want to have a look as you started this?

File: examples/simulation/plot_simulate_evoked_data.py [hunk 19]
```diff
@@ -37,44 +37,34 @@
...
-stc_data *= 100 * 1e-9  # use nAm as unit
-
-# time translation
-stc_data[1] = np.roll(stc_data[1], 80)
-stc = generate_sparse_stc(fwd['src'], labels, stc_data, tmin, tstep,
-                          random_state=0)
+# Generate source time courses from 2 dipoles and the correspond evoked data
+
+times = np.arange(300, dtype=np.float) / raw.info['sfreq'] - 0.1
+rng = np.random.RandomState(42)
```
Comment (wronk) [inline examples/simulation/plot_simulate_evoked_data.py]: Do we want to hard code this or leave the parameter as `None`?

File: mne/forward/forward.py [hunk 20]
```diff
@@ -1148,33 +1150,48 @@ def apply_forward(fwd, stc, evoked_template, start=None, stop=None,
...
-
     return evoked
 
 
+# XXX : apply_forward_raw should take info and not raw_template
+
+
 @verbose
 def apply_forward_raw(fwd, stc, raw_template, start=None, stop=None,
                       verbose=None):
```
Comment (yousrabk) [inline mne/forward/forward.py]: As the apply_forward, we should may be give the info instead of raw_template?

File: mne/simulation/evoked.py [hunk 21]
```diff
@@ -4,16 +4,59 @@
...
+
+    Parameters
+    ----------
+    fwd : dict
+        a forward solution.
+    stc : SourceEstimate object
+        The source time courses.
+    evoked : None | Evoked object
+        An instance of evoked used as template.
+    cov : Covariance object
```
Comment (yousrabk) [inline mne/simulation/evoked.py]: Sometimes it is Covariance object and sometimes it's instance of Covariance (same for evoked...), which one is to use?

File: mne/forward/forward.py (hunk 20 above)
Comment (agramfort) [inline mne/forward/forward.py]: +1

File: mne/simulation/evoked.py (hunk 21 above)
Comment (agramfort) [inline mne/simulation/evoked.py]: I would you instance of Covariance but we're not super consistent ...

File: mne/simulation/tests/test_evoked.py [hunk 22]
```diff
@@ -52,24 +49,14 @@ def test_simulate_evoked():
...
-    stc_data[1][:len(Ws[1])] = np.real(Ws[1])
-    stc_data *= 100 * 1e-9  # use nAm as unit
-
-    # time translation
-    stc_data[1] = np.roll(stc_data[1], 80)
-    stc = generate_sparse_stc(fwd['src'], labels, stc_data, tmin, tstep,
-                              random_state=0)
+    # Generate times series for 2 dipoles
+    stc = simulate_sparse_stc(fwd['src'], n_dipoles=2, times=times)
+    stc._data *= 1e-9
```
Comment (yousrabk) [inline mne/simulation/tests/test_evoked.py]: This should be removed as the *data_fun* in *simulate_sparse_stc* is already scaling the data

**Comment by yousrabk** [comment]
LGTM after updating the what's new. Thanks @Eric89GXL for taking over

**Comment by agramfort** [comment]
thanks @Eric89GXL @yousrabk @wronk 

I am taking care of what's new update in master

**Comment by Eric89GXL** [comment]
Thanks for the reviews, folks.

FYI @agramfort this week I plan on doing some preliminary/basic support for head position estimation from cHPI freqs. I'll keep the functions private for the release, but hopefully I'll have something workable enough to better validate the cHPI portion of the raw simulations. Eventually we'll want the estimation for doing movement compensation for Maxwell filtering, so it would be a good step to have.

**Comment by agramfort** [comment]
sounds good
